File: Technical/3 ClassNotes/Java/src/practiceProblems/SmallestSubarray.java
package practiceProblems;

public class SmallestSubarray {

    public static int minLength(int[] arr, int k) {

        int left = 0;
        int sum = 0;
        int ans = Integer.MAX_VALUE;

        for(int right = 0; right < arr.length; right++) {

            sum += arr[right];

            while(sum >= k) {

                ans = Math.min(ans,
                               right - left + 1);

                sum -= arr[left];
                left++;
            }
        }

        return ans == Integer.MAX_VALUE
                ? 0
                : ans;
    }

    public static void main(String[] args) {

        int[] arr = {2,3,1,2,4,3};

        System.out.println(minLength(arr, 7));
    }
}

File: Technical/3 ClassNotes/Python/practice_problems/smallest_subarray.py
def min_length(arr, k):

    left = 0
    current_sum = 0
    ans = float('inf')

    for right in range(len(arr)):

        current_sum += arr[right]

        while current_sum >= k:

            ans = min(ans,
                      right - left + 1)

            current_sum -= arr[left]
            left += 1

    return 0 if ans == float('inf') else ans


arr = [2,3,1,2,4,3]
print(min_length(arr, 7))